FAI-182: Score Cards: Add skeleton for PMML Score Cards (4 tabs in UX design)

manstis opened the pull request:
Hi @tiagobento this is my first hack at adding a TypeScript (PMML) editor to `kogito-tooling` channels (_online_ only, atm)

It's just the last commit that's important (the others relate to other JIRA/PR that this is built on top of.. they're unmerged atm).

Could you please give me your opinion?

I also need to know how to integrate `StateControl` with the mock/stub/test `PMMLEditor`, if you can advise?

manstis commented on packages/embedded-editor/src/envelope/envelope.ts:
```diff
@@ -14,22 +14,29 @@
 const gwtAppFormerApi = new GwtAppFormerApi();
 gwtAppFormerApi.setClientSideOnly(true);
 
+const editorFactories: Array<EditorFactory<LanguageData>> = new Array<EditorFactory<LanguageData>>();
```
TypeScript (or possibly `TSLint`) complains bitterly about using `[]` for _typed_ types (i.e. `EditorFactory<LanguageData>`) and use of `Array` is forced...

tiagobento replied:
Yeah, I think that's the default configuration for TSLint.

manstis commented on packages/microeditor-envelope/src/CompositeEditorFactory.ts:
```diff
@@ -0,0 +1,44 @@
+    languageData: LanguageData,
+    messageBus: EnvelopeBusInnerMessageHandler
+  ): Promise<AppFormer.Editor> {
+    const candidates: Array<EditorFactory<LanguageData>> = this.factories.filter(f => f.supports(languageData));
```
I lookup a _delegate_ factory based on the _type_ of the `LanguageData`.

tiagobento replied:
Nice! You don't need the explicit type declaration there though 😝

manstis replied:
OK.. updated but I shiver at the concept (as I am a Java/typed-language developer)!

manstis replied:
We'll need to think about what happens IF a factory cannot be found for a given `LanguageData`... we know it's not possible, but we don't protect against it (in the old AppFormer days we'd show a popup that an _activity could not be found_ -- it was loosely modelled on Android's _activities_).

tiagobento replied:
Yeah.. that sounds like a good feature for the Envelope itself. Showing a nice message saying that something went wrong or that the editor is not available for that specific file extension.

manstis commented on packages/microeditor-envelope/src/EditorFactory.ts:
```diff
@@ -22,6 +22,11 @@ import { EnvelopeBusInnerMessageHandler } from "./EnvelopeBusInnerMessageHandler
+   * Returns true if the factory supports a language.
+   * @param languageData The custom LanguageData
+   */
+  supports(languageData: T): boolean;
```
It was either to add this filtering of factory for type or have ALL factories attempt to build an editor and perform the filtering internally; returning an editor instance if they supported the type.. I thought this was cleaner.

tiagobento replied:
Agreed.

manstis commented on packages/online-editor/src/App.tsx:
```diff
@@ -47,6 +48,7 @@ export function App(props: Props) {
   const router: EmbeddedEditorRouter = useMemo(
     () =>
       new EmbeddedEditorRouter(
+        new PMMLRoute(),
```
New route for new file type.

manstis commented on packages/online-editor/src/home/HomePage.tsx:
```diff
@@ -607,7 +607,7 @@ export function HomePage(props: Props) {
                 {/* Transparent file input overlays the button */}
                 <input
-                  accept={".dmn, .bpmn, .bpmn2"}
+                  accept={".dmn, .bpmn, .bpmn2., .pmml"}
```
Allow me to upload a local file to test.

tiagobento replied:
I'm assuming this is a temporary change so that you can test the PMML editor on the Online channel. Ideally we would have a rework on the Home screen if we decide to have the PMML editor on the Online channel as well, right?

manstis replied:
Absolutely... ALL the changes in `online-editor` are just to showcase/demo the `CompositeEditorFactory`. When I say ALL, it's up for debate.. I _could_ keep the `CompositeEditorFactory` in the PR but don't register the `PMMLEditorFactory` or this change or the additional route...

tiagobento replied:
I like that the `CompositeEditorFactory` is in microeditor-envelope. People creating envelopes *will* have access to all the modules providing editors (along with their respective factories), so having the `CompositeEditorFactory` class there is nice.

manstis commented on packages/online-editor/static/images/pmml_kogito_logo.svg:
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Generator: Adobe Illustrator 23.1.1, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
+<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
```
Copy of DMN's SVG logo for the `EditorHeader`.. PMML shows as "DMN" for now....

manstis commented on packages/pmml-editor/src/editor/PMMLEditor.tsx:
```diff
@@ -0,0 +1,79 @@
+        return new Promise<void>(resolve => {
+            this.setState({ content: content }, () => {
+                console.log("Setting content to " + content);
+                this.props.messageBus.notify_dirtyIndicatorChange(this.isDirty());
```
This mechanism seems obsolete. I'll need to integrate `StateControl`.. pointers welcome :-)

tiagobento replied:
Yes.. it is. I'll remove this message on a future PR.

tiagobento replied:
You'll want to update that once #200 is merged.

tiagobento replied:
I think the whole callback is not necessary anymore.

manstis commented:
The last commit also moves a few bits around (only PMML stuff):-
- `pmml-editor-codegen` - unchanged; this is where I generate the TS classes from the PMML XSD/Java classes
- `pmml-editor-marshaller` - now only contains the marshaller (moved to a separate module as DMN has a need for just this marshaller to support PMML documents client-side; and trusty-ai needs it too to show a viewer for PMML documents - short term, will be moving to `kogito-tooling`).
- `pmml-editor` - the editor itself (stolen from [here](https://github.com/kiegroup/kogito-tooling-examples/tree/master/packages/simple-react-editors/src)!)

tiagobento commented:
@manstis Michael, can you please update this draft after #189 is merged? I guess it'll be so much easier for me to review :)

manstis commented:
@tiagobento Absolutely... I have another PR in the pipeline first though (but not submitted until #189 is merged too)

manstis commented:
@tiagobento This is ready for you to assassinate now :-)

The changes to `online-editor` are there to demonstrate _how_ additional types of editor _could_ be supported. I suspect the changes to `online-editor` themselves WILL NOT be merged in this PR (or at least the specifics for PMML will be removed). When you're happy with the approach I'll remove the PMML specific bits in `online-editor` and write tests where needed.

tiagobento commented on packages/microeditor-envelope/src/CompositeEditorFactory.ts:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  public supports(languageData: LanguageData) {
+    return true;
```
Shouldn't you check if the provided factories support the argument languageData too?

manstis replied:
Updated.

tiagobento commented on packages/pmml-editor-marshaller/package.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "@kogito-tooling/pmml-editor-marshaller",
+  "version": "0.5.2",
+  "description": "",
+  "license": "Apache-2.0",
+  "main": "./dist/index.js",
+  "types": "./dist/index.d.ts",
+  "files": [
+    "dist"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/kiegroup/kogito-tooling.git"
+  },
+  "dependencies": {
+    "jsonata": "^1.8.3",
+    "xml-js": "^1.6.11"
+  },
+  "scripts": {
+    "lint": "tslint -c ../../tslint.json 'src/**/*.{ts,tsx,js,jsx}'",
+    "test": "jest",
+    "build:fast": "rm -rf dist && webpack",
+    "build": "yarn run lint && yarn test && yarn run build:fast",
+    "build:prod": "yarn run build --mode production --devtool none"
+  },
+  "babel": {
+    "presets": [
+      "@babel/env"
+    ]
+  }
+}
```
Thanks for that. I'm planning a rework on our package names and organisation and this really aligns well with it. I'll provide more details soon :D

manstis replied:
... and is essential for the bigger world!

- The DMN Editor in `kie-wb-common-dmn` needs a client-side PMML marshaller
- The Linear Regression Viewer in `trusty-ui-sandbox` also needs a PMML marshaller

The _plan_ is for both to reference this `pmml-marshaller` package.....

tiagobento replied:
Nice! :)

manstis commented:
Hi @tiagobento what/where/how etc are the changes you're making going to affect this?

I could do with the `pmml` package split (`pmml-marshaller`, `pmml-editor`) in `kogito-tooling` soon as the marshaller is needed elsewhere. If you think the impact on the changes I've made for React-based editors are going to be extensive I might as well split this PR into two; one adding the split for `pmml` packages - to be merged sooner that later - and the other PR worrying about a basic PMML editor implementation (wired into the `StateControl`, the new `Promise` work etc).

If you think the impact here is manageable and coming soon I'll sit tight...

manstis commented:
@tiagobento Hello, this is "good to go" now...

I've removed the example use in `online-editor` and added unit tests.

I expect this PR to change before it gets to the front of the queue due to other PRs in the pipeline. 

@jomarko Given the above comment; feel free to leave any review for now, until @tiagobento has approved, other PRs in front of this in the queue have been merged and any fallout changes completed here.

tiagobento commented on packages/embedded-editor/src/envelope/envelope.ts:
```diff
@@ -14,19 +14,24 @@
+import { CompositeEditorFactory, EditorFactory } from "@kogito-tooling/microeditor-envelope";
 import { EnvelopeBusMessage } from "@kogito-tooling/microeditor-envelope-protocol";
 
+const editorFactories: Array<EditorFactory<LanguageData>> = new Array<EditorFactory<LanguageData>>();
+editorFactories.push(new GwtEditorWrapperFactory());
+const compositeEditorFactory: CompositeEditorFactory = new CompositeEditorFactory(editorFactories);
```
```suggestion
const compositeEditorFactory = new CompositeEditorFactory([new GwtEditorWrapperFactory()]);
```

manstis replied:
Done and inlined.

tiagobento commented on packages/embedded-editor/src/envelope/envelope.ts:
```diff
@@ -14,19 +14,24 @@
     }
   },
-  editorFactory: new GwtEditorWrapperFactory(),
+  editorFactory: compositeEditorFactory,
```
You can even inline this variable here.

tiagobento commented on packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts:
```diff
@@ -32,13 +32,19 @@ declare global {
   }
 }
 
+export const FACTORY_TYPE: string = "gwt";
```
You can drop the explicit type definition.

manstis replied:
Oeeeee... well.... I _like_ using TS's type system :-)

Should we not be trying to use it more extensively? Is that not one of the reasons for using TS (over JS)?

Is this where JS vs Java developers worlds collide!! I hated writing JS and having to guess parameters types or stepping through implementations to find whether it was expected to be a primitive or object or either when documentation was missing...

tiagobento replied:
Yeah..... well, Java has `var` now too, which does basically the same type-inference mechanism :P

In TypeScript, string literals are types too, and since you're creating a `const` variable, the inferred type of `FACTORY_TYPE` is actually `"gwt"`, and not `string`. Sometimes I end up writing the types explicitly too, but only when I need to make something not ambiguous. 

Specially when I'm creating object literals, sometimes you want them to be of a specific type/interface, instead of the inferred. Interesting is that TypeScript matches types "by value", so two separate types/interfaces defining the exactly same properties are actually compatible.

One thing that helps, is on your IDE, you can probably hover over a variable/function to see its explicitly-typed signature. On WebStorm I get this:

![image](https://user-images.githubusercontent.com/1584568/87956937-02d28180-ca86-11ea-8c96-fecae7675952.png)

This is more like a "cultural shock" between Java and TypeScript worlds, but TypeScript actually does a pretty decent job with its type-inference. I know it might look weird to you, but I believe this kind of stuff makes your code more idiomatic to other TypeScript developers and it's something you'll get used to read and write with the time :)

I don't think this is a requirement for me to approve this nor future PRs, since the code is actually correct and there are no major impacts. I know it's a lot to take in moving from Java to TypeScript, but I'd really like for us to try and "dive in" the TypeScript-way of writing code. Hope that makes sense you!

manstis replied:
OK, good to know.. I want to follow the _best practice_ for TS development.

Updated: https://github.com/kiegroup/kogito-tooling/pull/197/commits/2e14fcc651c8b5391adc1bc332ff35de95557365

I refer to Typescript's type/interface matching as "shape"... if the two JSON objects have the same structure they're the same... it works to our advantage nicely many times.

tiagobento commented on packages/microeditor-envelope/src/CompositeEditorFactory.ts:
```diff
@@ -0,0 +1,44 @@
+ * to specific concrete factories. The first concrete factory to return a promise to create an editor is selected.
+ */
+export class CompositeEditorFactory implements EditorFactory<LanguageData> {
+  private factories: Array<EditorFactory<LanguageData>> = new Array<EditorFactory<LanguageData>>();
+
+  constructor(factories: Array<EditorFactory<LanguageData>>) {
+    this.factories = factories;
+  }
```
```suggestion
  constructor(private readonly factories: Array<EditorFactory<LanguageData>>) {
  }
```

manstis replied:
Wow. That is some black-art voodoo syntax (that this ol' Java-boy will have to learn!)

tiagobento commented on packages/microeditor-envelope/src/CompositeEditorFactory.ts:
```diff
@@ -0,0 +1,44 @@
+    messageBus: EnvelopeBusInnerMessageHandler
+  ): Promise<AppFormer.Editor> {
+    const candidates = this.factories.filter(f => f.supports(languageData));
+    return candidates[0].createEditor(languageData, messageBus);
```
Can you throw an error if `candidates.length !== 1`?

manstis replied:
Sure thing done... and test added!

tiagobento commented on packages/microeditor-envelope/src/CompositeEditorFactory.ts:
```diff
@@ -0,0 +1,44 @@
+    return this.factories.filter(f => f.supports(languageData)).length > 0;
+  }
+
+  public createEditor(
+    languageData: LanguageData,
+    messageBus: EnvelopeBusInnerMessageHandler
```
The signature of this method changed a bit on #205, just a heads up since it's before in the merge queue.

tiagobento commented on packages/pmml-editor/src/editor/PMMLEditor.tsx:
```diff
@@ -0,0 +1,79 @@
+        return this.setContent(this.state.path, this.state.content).then(() => this.state.content);
+    }
+
+    public isDirty() {
+        return this.state.content !== this.state.originalContent;
+    }
```
This method can be deleted.

manstis replied:
Done.

tiagobento commented on packages/pmml-editor/src/editor/PMMLEditorInterface.tsx:
```diff
@@ -0,0 +1,59 @@
+        return Promise.resolve(undefined);
+    }
+
+    public isDirty(): boolean {
+        return this.self.isDirty();
+    }
```
This can be removed.

manstis replied:
Done.

tiagobento commented on packages/pmml-editor/src/editor/PMMLEditorInterface.tsx:
```diff
@@ -0,0 +1,59 @@
+        return <PMMLEditor exposing={s => (this.self = s)} messageBus={this.messageBus} />;
+    }
+
+    public async undo(): Promise<void> {
+        //Place holder until StateControl is added for React based components.
+    }
+
+    public async redo(): Promise<void> {
+        //Place holder until StateControl is added for React based components.
+    }
```
Actually, that's it. You can create undo and redo methods on your PMMLEditor component that know how to undo and redo changes.

tiagobento replied:
No need to change that in this PR, just FYI.

manstis replied:
OK, the `PMMLEditor` class is just a place holder at the moment. It was demo'ed briefly in the `online-editor` to prove different editor factories were being matched and used.. but since that code has been reverted this class now simple marks _hook_ for the real editor (that will live in `pmml-editor`...).

manstis commented:
@tiagobento Updates made (other than [this](https://github.com/kiegroup/kogito-tooling/pull/197#discussion_r455902986) one as IDK where/what you exactly mean).

I'll be updating this after #205 so can make the change then... if you can point it out to me.

tiagobento commented on packages/kie-bc-editors/src/GwtEditorWrapperFactory.ts:
```diff
@@ -32,13 +32,19 @@ declare global {
   }
 }
 
+export const FACTORY_TYPE: string = "gwt";
```
```suggestion
export const FACTORY_TYPE = "gwt";
```

That's what I meant :D

manstis replied:
Done.

tiagobento commented:
@manstis Thanks! Just added a comment suggestion there :)

tiagobento commented on packages/online-editor/src/App.tsx:
```diff
@@ -23,6 +23,7 @@ import * as React from "react";
 import { useCallback, useMemo, useState } from "react";
 import { Route, Switch } from "react-router";
 import { HashRouter } from "react-router-dom";
+import "../static/resources/style.css";
```
Last one: Can you revert the changes on this file? We've had problems with CSS-loading order before and before we investigate that in more detail I don't wanna run any risks.

manstis replied:
Done.

jomarko commented:
@manstis , hi please let me know once this has conflicts resolved

manstis commented:
@jomarko @ederign This is ready for your review.

@tiagobento IDK if you want to re-review as some changes were needed after rebasing with `master`.

tiagobento commented:
Hi @manstis, thanks for updating this PR. Can you please check your `webpack.config.js` files? I've created `common` file so that modules can define shorter configuration files. You can update your to use that.

manstis commented:
@tiagobento I've updated the `webpack.config.js` file in a new commit.

ederign commented:
@jomarko do you have any estimative when you will gonna be able to review this?

manstis commented:
@ederign @jomarko is on PTO.
